lib/internal/fs/promises.js

```diff
@@ -792,48 +794,72 @@ async function readdirRecursive(originalPath, options) {
       ),
     ],
   ];
-
+  function filter(absolutePath) {
+    const directory = absolutePath.slice(0, absolutePath.lastIndexOf(sep));
+    return absolutePath === originalPath ||
+      typeof options.recursiveFilter !== 'function' ||
+      (typeof options.recursiveFilter === 'function' && options.recursiveFilter({
+        depth: absolutePath.match(lengthReg).length - baseDirLength,
+        itemName: absolutePath.slice(absolutePath.lastIndexOf(sep) + 1),
+        itemPath: absolutePath,
+        parentName: directory.slice(directory.lastIndexOf(sep) + 1),
+        parentPath: directory,
+      }));
+  }
+  const sep = pathModule.sep;
+  const lengthReg = (sep === '\\') ?
+    new RegExp('\\\\', 'g') :
+    new RegExp('/', 'g');
+  const baseDirLength = originalPath.match(lengthReg).length;
 
   if (options.withFileTypes) {
     while (queue.length > 0) {
       // If we want to implement BFS make this a `shift` call instead of `pop`
       const { 0: path, 1: readdir } = ArrayPrototypePop(queue);
-      for (const dirent of getDirents(path, readdir)) {
-        ArrayPrototypePush(result, dirent);
-        if (dirent.isDirectory()) {
-          const direntPath = pathModule.join(path, dirent.name);
-          ArrayPrototypePush(queue, [
-            direntPath,
-            await binding.readdir(
-              direntPath,
-              options.encoding,
-              true,
-              kUsePromises,
-            ),
-          ]);
+      if (filter(pathModule.resolve(path))) {
+        for (const dirent of getDirents(path, readdir)) {
+          if (filter(dirent)) {
```

Comment:
It seems that this code will call `filter()` twice for processed directories.

Reply:
The first time filters for directory traversal and the second time is for items in the given directory.

Reply:
Which means, every subdirectory will be passed to `filter()` as item of its parent, and then again as current directory.

IMHO, one filter here
```diff
-if (dirent.isDirectory()) {
+if (dirent.isDirectory() && filter(dirent, depth)) {
```
would be sufficient in case if we want to be able to prevent further `readdir` calls by returning falsey value, and returning regular files unconditionally.